Any plan to support node ESM?

## I'm submitting a...
<!-- Check one of the following options with "x" -->

 - [ ] Bug report  <!-- Please search GitHub for a similar issue or PR before submitting -->
 - [x] Feature request
 - [ ] Question

## Current behavior
The default generated migrations are all CJS files and can not be used in a ESM node project (`type="module"` in package.json).

It seems that node-db-migrate hardcoded the `.js` suffix. If I change the migration files extension to `.cjs`, it can't recognize them at all.

## Expected behavior
Provide an option to generate ESM migration code or recognize `.cjs` migration files.

## Environment

<pre><code>
db-migrate version: 1.0.0-beat.14
plugins with versions: none
db-migrate driver with versions: sqlite3 0.4.0

Additional information:
- Node version: 15.5.0
- Platform:  Mac

I am sorry to be ambiguous, let me clarify this.

I have a project, and I set it up to be an ESM project, that is the `type` field in package.json is `module`.

I use the `migrations` directory the store the auto-generated migration files.

When I do the migration, I got this error:

```
[ERROR] Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: /Users/cj/Documents/Work/pdb-api/migrations/20210110045153-init.js
require() of ES modules is not supported.
require() of /Users/cj/Documents/Work/pdb-api/migrations/20210110045153-init.js from /Users/cj/Documents/Work/pdb-api/node_modules/db-migrate/lib/file.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
Instead rename 20210110045153-init.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /Users/cj/Documents/Work/pdb-api/package.json.
```

The reason is that node-db-migrate is trying to `require` the migration file but the file is considered to be an ESM module (due to the `type` filed in the root package.json file). Obviously, one can't load an ESM module using `require`.

Back then I was busy writing the code so I didn't give too much time to think about it but filed this issue.

For future references, there are several ways to solve this problem without any modification of node-db-migrate.

1. Create a `package.json` file in the `migrations` directory and set its `type` filed to be anything except `module`. This is probably the easiest way.
2. Remove `type="module"` in root package.json and manually change all file extensions of our code from `.js` to `.mjs`.

--- a/lib/commands/create-migration.js
+++ b/lib/commands/create-migration.js
@@ -5,18 +5,38 @@ const log = require('db-migrate-shared').log;
 const mkdirp = Promise.promisify(require('mkdirp'));
 const fs = require('fs');
 const stat = Promise.promisify(fs.stat);
+const writeFile = Promise.promisify(fs.writeFile);
 const yargs = require('yargs');
 const util = require('util');
+const path = require('path');
 
 async function createMigrationDir (dir) {
   const res = await stat(dir).catch(_ => {
     return { err: true };
   });
   if (res && res.err === true) {
-    return mkdirp(dir);
+    await mkdirp(dir);
   }
 
-  return Promise.resolve();
+  // Create migrations/package.json to ensure migration files are
+  // executed as CJS and not ESM
+  // https://github.com/db-migrate/node-db-migrate/issues/721
+  const packageJsonPath = path.resolve(dir, 'package.json');
+
+  await stat(packageJsonPath).then(
+    async () => {
+      const packageJson = require(packageJsonPath);
+      packageJson.type = 'commonjs';
+      const packageJsonStr = JSON.stringify(packageJson, null, 2);
+      await writeFile(packageJsonPath, packageJsonStr, 'utf-8');
+    },
+    async (err) => {
+      const packageJson = JSON.stringify({
+        type: 'commonjs'
+      }, null, 2)
+      await writeFile(packageJsonPath, packageJson, 'utf-8');
+    }
+  )
 }
 
 async function executeCreateMigration (internals, config) {

--- a/test/integration/create_test.js
+++ b/test/integration/create_test.js
@@ -8,8 +8,10 @@ const fs = require('fs');
 const path = require('path');
 const cp = require('child_process');
 const dbmUtil = require('db-migrate-shared').util;
+const mkdirp = Promise.promisify(require('mkdirp'));
 
 const rmdir = Promise.promisify(require('rimraf'));
+const writeFile = Promise.promisify(fs.writeFile);
 
 function wipeMigrations () {
   const dir = path.join(__dirname, 'migrations');
@@ -49,9 +51,48 @@ lab.experiment('create', function () {
 
     lab.test('will create a new migration', () => {
       const files = fs.readdirSync(path.join(__dirname, 'migrations'));
-      Code.expect(files.length).to.equal(1);
-      const file = files[0];
-      Code.expect(file).to.match(/first-migration\.js$/);
+      const file = files.find(file => /first-migration\.js$/.test(file));
+      Code.expect(file).to.exist();
+    });
+
+    lab.test('will create migrations/package.json', () => {
+      const files = fs.readdirSync(path.join(__dirname, 'migrations'));
+      const packageJson = files.find(file => /package\.json$/.test(file));
+      Code.expect(packageJson).to.exist();
+    })
+  });
+
+  lab.experiment(
+    'with existing migrations/package.json file', () => {
+    let exitCode;
+
+    lab.before(async () => {
+      await wipeMigrations();
+
+      await mkdirp(path.join(__dirname, 'migrations'));
+
+      await writeFile(path.join(__dirname, 'migrations', 'package.json'), '{"name": "test", "type": "module"}');
+
+      const db = dbMigrate('create', 'first migration');
+      // db.stderr.on('data', data => console.log(data.toString()));
+      // db.stdout.on('data', data => console.log(data.toString()));
+
+      const exitCodePromise = new Promise((resolve) => db.on('exit', resolve));
+      exitCode = await exitCodePromise;
+    });
+
+    lab.test('does not cause an error', () => {
+      Code.expect(exitCode).to.equal(0);
+    });
+
+    lab.test('will modify an existing migrations/package.json\'s type field', () => {
+      const packageJson = JSON.parse(fs.readFileSync(path.join(__dirname, 'migrations', 'package.json')))
+      Code.expect(packageJson.type).to.equal("commonjs");
+    });
+
+    lab.test('will preserve other properties in an existing package.json', () => {
+      const packageJson = JSON.parse(fs.readFileSync(path.join(__dirname, 'migrations', 'package.json')))
+      Code.expect(packageJson.name).to.equal("test");
     });
   });
 
@@ -84,7 +125,7 @@ lab.experiment('create', function () {
         for (let i = 0; i < files.length; i++) {
           const file = files[i];
           const stats = fs.statSync(path.join(__dirname, 'migrations', file));
-          if (stats.isFile()) {
+          if (stats.isFile() && !/package\.json$/.test(file)) {
             Code.expect(file).to.match(/second-migration\.js$/);
           }
         }
@@ -97,9 +138,8 @@ lab.experiment('create', function () {
 
       lab.test('will create a new migration sql up file', () => {
         const files = fs.readdirSync(path.join(__dirname, 'migrations/sqls'));
-        Code.expect(files.length).to.equal(2);
-        const file = files[1];
-        Code.expect(file).to.match(/second-migration-up\.sql$/);
+        const file = files.find(file => /second-migration-up\.sql$/.test(file));
+        Code.expect(file).to.exist();
       });
     }
   );
@@ -129,7 +169,7 @@ lab.experiment('create', function () {
         for (let i = 0; i < files.length; i++) {
           const file = files[i];
           const stats = fs.statSync(path.join(__dirname, 'migrations', file));
-          if (stats.isFile()) {
+          if (stats.isFile() && !/package\.json$/.test(file)) {
             Code.expect(file).to.match(/third-migration\.js$/);
           }
         }
@@ -142,9 +182,8 @@ lab.experiment('create', function () {
 
       lab.test('will create a new migration sql up file', () => {
         const files = fs.readdirSync(path.join(__dirname, 'migrations/sqls'));
-        Code.expect(files.length).to.equal(2);
-        const file = files[1];
-        Code.expect(file).to.match(/third-migration-up\.sql$/);
+        const file = files.find(file => /third-migration-up\.sql$/.test(file));
+        Code.expect(file).to.exist();
       });
     }
   );
@@ -177,7 +216,7 @@ lab.experiment('create', function () {
         for (let i = 0; i < files.length; i++) {
           const file = files[i];
           const stats = fs.statSync(path.join(__dirname, 'migrations', file));
-          if (stats.isFile()) {
+          if (stats.isFile() && !/package\.json$/.test(file)) {
             Code.expect(file).to.match(/fourth-migration\.coffee$/);
           }
         }
@@ -211,9 +250,8 @@ lab.experiment('create', function () {
 
         lab.test('will create a new migration', () => {
           const files = fs.readdirSync(path.join(__dirname, 'migrations/test'));
-          Code.expect(files.length).to.equal(2);
-          const file = files[0];
-          Code.expect(file).to.match(/first-migration\.js$/);
+          const file = files.find(file => /first-migration\.js$/.test(file));
+          Code.expect(file).to.exist();
         });
         lab.test('will create a new migration/test/sqls directory', () => {
           const stats = fs.statSync(path.join(__dirname, 'migrations/test/sqls'));
@@ -221,9 +259,8 @@ lab.experiment('create', function () {
         });
         lab.test('will create a new migration sql up file', () => {
           const files = fs.readdirSync(path.join(__dirname, 'migrations/test/sqls'));
-          Code.expect(files.length).to.equal(2);
-          const file = files[1];
-          Code.expect(file).to.match(/first-migration-up\.sql$/);
+          const file = files.find(file => /first-migration-up\.sql$/.test(file));
+          Code.expect(file).to.exist();
         });
       });
     }
@@ -254,7 +291,7 @@ lab.experiment('create', function () {
         for (let i = 0; i < files.length; i++) {
           const file = files[i];
           const stats = fs.statSync(path.join(__dirname, 'migrations', file));
-          if (stats.isFile()) {
+          if (stats.isFile() && !/package\.json$/.test(file)) {
             Code.expect(file).to.match(/fifth-migration\.coffee$/);
           }
         }
@@ -306,7 +343,7 @@ lab.experiment('create', function () {
         for (let i = 0; i < files.length; i++) {
           const file = files[i];
           const stats = fs.statSync(path.join(__dirname, 'migrations', file));
-          if (stats.isFile()) {
+          if (stats.isFile() && !/package\.json$/.test(file)) {
             Code.expect(file).to.match(/sixth-migration\.js$/);
           }
         }

